File: TheMartianChronicles/graph_bfs/Leet102.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 * 76% time percentile, 52 ms, 35 mb
 */
var levelOrder = function(root) {
  if(!root){
    return [];
  }
  var result = [];
  var queue = [];
  
  queue.push(root);
  // let levels = 0;
  
  while(queue.length) {
    
    let level = queue.length;
    
    let temp = [];
    for(let i = 0; i < level; i++) {
      let currentNode = queue.shift();
      temp.push(currentNode.val);
       //unlink and then add the neighbors from the queue
      if(currentNode.left !==null) {
        queue.push(currentNode.left); 
      }
      if(currentNode.right !== null) {
        queue.push(currentNode.right);
      } 
      
      //next level...avicii
    }
    
    result.push(temp);//add current array on the current level for breadth first searching
  }
  
  return result;
  
};

File: TheMartianChronicles/graph_bfs/Leet102.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root is None: return []
        q = [[root]]
        for level in q:
          record = []
          for node in level:
            if node.left: record.append(node.left)
            if node.right: record.append(node.right)
          if record: q.append(record)
        return [[x.val for x in level] for level in q]